### PR TITLE
Added ensure funcs for serviceaccounts and clusterroles 

### PR DESCRIPTION
**Pull request by thejasn**
RFE: [CFE-316](https://issues.redhat.com/browse/CFE-316)

Signed-off-by: thejasn <thn@redhat.com>

### CODE REVIEW

File: config/rbac/role.yaml [hunk 1]
```diff
@@ -98,8 +110,10 @@ rules:
   - rolebindings
   - roles
   verbs:
+  - bind
```
Comment (arjunrn) [inline config/rbac/role.yaml:113]: Let's not use `bind` and `escalate`. Instead let's grant the operator all the permissions which the controller needs. The `bind` and `escalate` approach will not pass security.

File: hack/sync-upstream-crds.sh [hunk 2]
```diff
@@ -1,7 +1,7 @@
 #!/bin/bash
 set -euo pipefail
 
-UPSTREAM="https://raw.githubusercontent.com/openshift/aws-load-balancer-controller/main/config/crd/bases"
+UPSTREAM="https://raw.githubusercontent.com/kubernetes-sigs/aws-load-balancer-controller/main/config/crd/bases"
```
Comment (arjunrn) [inline hack/sync-upstream-crds.sh:4]: Why was this changed?

File: pkg/controllers/awsloadbalancercontroller/rbac_clusterrole.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	albo "github.com/openshift/aws-load-balancer-operator/api/v1alpha1"
+)
+
+// ensureClusterRoles creates and maintains clusterroles required by the aws-load-balancer controller.
+func (r *AWSLoadBalancerControllerReconciler) ensureClusterRoles(ctx context.Context, controller *albo.AWSLoadBalancerController) error {
```
Comment (arjunrn) [inline pkg/controllers/awsloadbalancercontroller/rbac_clusterrole.go:19]: Let's change this and the corresponding functions to behave like there is a single `ClusterRole`.

File: pkg/controllers/awsloadbalancercontroller/rbac.go [hunk 4]
```diff
@@ -3,9 +3,93 @@ package awsloadbalancercontroller
...
+
+	err = r.ensureRoleBindings(ctx, sa, controller)
+	if err != nil {
+		return err
+	}
+
 	return nil
 }
+
+func buildClusterRole(name string, rules []rbacv1.PolicyRule) *rbacv1.ClusterRole {
```
Comment (arjunrn) [inline pkg/controllers/awsloadbalancercontroller/rbac.go:37]: Could you put these _build*_ functions in the files where they are used.

File: pkg/controllers/awsloadbalancercontroller/rbac_clusterrole.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+	exist, current, err := r.currentClusterRoles(ctx, desired)
+	if err != nil {
+		return fmt.Errorf("failed to fetch current clusterroles: %w", err)
+	}
+
+	if !exist {
+		if err := r.createClusterRoles(ctx, desired); err != nil {
+			return err
+		}
+		_, current, err = r.currentClusterRoles(ctx, desired)
```
Comment (arjunrn) [inline pkg/controllers/awsloadbalancercontroller/rbac_clusterrole.go:39]: You don't need to get the cluster role again after creating it because it's not passed back to the main controller reconciler loop.

File: pkg/controllers/awsloadbalancercontroller/rbac_clusterrole.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+	} else if updated {
+		_, _, err = r.currentClusterRoles(ctx, desired)
+		return err
+	}
+
+	return nil
+}
+
+// updateClusterRoles updates the current clusterroles and returns a flag to denote if the update was done.
+func (r *AWSLoadBalancerControllerReconciler) updateClusterRoles(ctx context.Context, current map[string]*rbacv1.ClusterRole, desired []*rbacv1.ClusterRole) (bool, error) {
```
Comment (arjunrn) [inline pkg/controllers/awsloadbalancercontroller/rbac_clusterrole.go:57]: The _updated_ flag is not required because the re-read in the calling function is really not required.

File: pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+	if !(reflect.DeepEqual(current.Subjects, desired.Subjects)) || !(reflect.DeepEqual(current.RoleRef, desired.RoleRef)) {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (r *AWSLoadBalancerControllerReconciler) createClusterRoleBindings(ctx context.Context, roles []*rbacv1.ClusterRoleBinding) error {
+	for _, role := range roles {
+		fmt.Printf("%+v\n", role)
```
Comment (arjunrn) [inline pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go:104]: Debug statement

File: pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+			return fmt.Errorf("failed to create clusterrolebinding %s/%s: %w", role.Namespace, role.Name, err)
+		}
+	}
+	return nil
+}
+
+func (r *AWSLoadBalancerControllerReconciler) currentClusterRoleBindings(ctx context.Context, roles []*rbacv1.ClusterRoleBinding) (bool, map[string]*rbacv1.ClusterRoleBinding, error) {
+	current := make(map[string]*rbacv1.ClusterRoleBinding)
+	for _, i := range roles {
+		fmt.Println(i)
```
Comment (arjunrn) [inline pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go:118]: Another debug statement.

File: pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+			}
+			return fmt.Errorf("failed to create clusterrolebinding %s/%s: %w", role.Namespace, role.Name, err)
+		}
+	}
+	return nil
+}
+
+func (r *AWSLoadBalancerControllerReconciler) currentClusterRoleBindings(ctx context.Context, roles []*rbacv1.ClusterRoleBinding) (bool, map[string]*rbacv1.ClusterRoleBinding, error) {
+	current := make(map[string]*rbacv1.ClusterRoleBinding)
+	for _, i := range roles {
```
Comment (arjunrn) [inline pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go:117]: nit: `i` suggests this variable is an index.

File: pkg/controllers/awsloadbalancercontroller/controller.go [hunk 10]
```diff
@@ -44,6 +44,10 @@ const (
 	controllerName               = "cluster"
 	controllerSecretName         = "albo-cluster-credentials"
 	controllerServiceAccountName = "albo-cluster-sa"
+	controllerClusterRole        = "albo-cluster-clusterrole"
```
Comment (arjunrn) [inline pkg/controllers/awsloadbalancercontroller/controller.go:47]: Actually the name of the resources should be `aws-load-balancer-controller-$INSTANCE`. You don't need the type in the name because the type is already known.

**Review by arjunrn — COMMENTED** [review]
Overall looks good, let's refactor this so that only one `ClusterRole`, `ClusterRoleBinding`, `Role` and `RoleBinding` are ensured and their name is generated with the instance name.

File: pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go (hunk 7 above)
Comment (thejasn) [reply pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go:104]: Wasn't required, removed it.

File: pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go (hunk 8 above)
Comment (thejasn) [reply pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go:118]: Wasn't required, removed it.

File: pkg/controllers/awsloadbalancercontroller/rbac_clusterrole.go (hunk 5 above)
Comment (thejasn) [reply pkg/controllers/awsloadbalancercontroller/rbac_clusterrole.go:39]: Addressed

File: pkg/controllers/awsloadbalancercontroller/rbac.go (hunk 4 above)
Comment (thejasn) [reply pkg/controllers/awsloadbalancercontroller/rbac.go:37]: Addressed

File: pkg/controllers/awsloadbalancercontroller/rbac_clusterrole.go (hunk 3 above)
Comment (thejasn) [reply pkg/controllers/awsloadbalancercontroller/rbac_clusterrole.go:19]: Addressed

File: hack/sync-upstream-crds.sh (hunk 2 above)
Comment (thejasn) [reply hack/sync-upstream-crds.sh:4]: Reverted.

File: config/rbac/role.yaml (hunk 1 above)
Comment (thejasn) [reply config/rbac/role.yaml:113]: Addressed

File: pkg/controllers/awsloadbalancercontroller/controller.go (hunk 10 above)
Comment (thejasn) [reply pkg/controllers/awsloadbalancercontroller/controller.go:47]: Since all resource names converge on a single name, I'll just create a single const (generic resource name), maybe later on we can add something like [this](https://github.com/openshift/external-dns-operator/blob/main/pkg/operator/controller/names.go)?

File: config/rbac/role.yaml (hunk 1 above)
Comment (arjunrn) [reply config/rbac/role.yaml:213]: You didn't actually remove these permissions from the file.

File: pkg/controllers/awsloadbalancercontroller/rbac_clusterrole.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+	updated.Rules = desired.Rules
+
+	if err := r.Client.Update(ctx, updated); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func isClusterRoleChanged(current *rbacv1.ClusterRole, desired *rbacv1.ClusterRole) (bool, error) {
```
Comment (arjunrn) [inline pkg/controllers/awsloadbalancercontroller/rbac_clusterrole.go:77]: `hasClusterRoleChanged` sounds like a better name.

File: pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+	updated.RoleRef = desired.RoleRef
+
+	if err := r.Client.Update(ctx, updated); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func isClusterRoleBindingChanged(current *rbacv1.ClusterRoleBinding, desired *rbacv1.ClusterRoleBinding) (bool, error) {
```
Comment (arjunrn) [inline pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go:78]: `hasClusterRoleBindingChanged` sounds like a better name.

Comment (arjunrn) [reply pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go:78]: This function does not need to return an error. I don't see an error returned anywhere.

File: pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+	if current.Name != desired.Name {
+		return true, nil
+	}
+
+	if current.Namespace != desired.Namespace {
+		return true, nil
+	}
```
Comment (arjunrn) [inline pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go:85]: Comparing the name and namespace is not really required because the current is retrieved from the name and the namespace of the desired.

File: pkg/controllers/awsloadbalancercontroller/rbac_role.go [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+
+func isRoleChanged(current *rbacv1.Role, desired *rbacv1.Role) (bool, error) {
+	if current.Name != desired.Name {
+		return true, nil
+	}
+
+	if current.Namespace != desired.Namespace {
+		return true, nil
+	}
+
```
Comment (arjunrn) [inline pkg/controllers/awsloadbalancercontroller/rbac_role.go:84]: Comparing name and namespace is not required.

File: pkg/controllers/awsloadbalancercontroller/rbac_role.go [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+	updated.Rules = desired.Rules
+
+	if err := r.Client.Update(ctx, updated); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func isRoleChanged(current *rbacv1.Role, desired *rbacv1.Role) (bool, error) {
```
Comment (arjunrn) [inline pkg/controllers/awsloadbalancercontroller/rbac_role.go:76]: `hasRoleChanged`

File: pkg/controllers/awsloadbalancercontroller/rbac_rolebinding.go [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+	updated.RoleRef = desired.RoleRef
+
+	if err := r.Client.Update(ctx, updated); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func isRoleBindingChanged(current *rbacv1.RoleBinding, desired *rbacv1.RoleBinding) (bool, error) {
```
[the remark this answers is not in the document]
Comment (arjunrn) [reply pkg/controllers/awsloadbalancercontroller/rbac_rolebinding.go:77]: Also drop the return error.

File: pkg/controllers/awsloadbalancercontroller/rbac_role.go (hunk 15 above)
Comment (arjunrn) [reply pkg/controllers/awsloadbalancercontroller/rbac_role.go:76]: Also drop the return error.

File: pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go (hunk 12 above)
Comment (thejasn) [reply pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go:78]: Good catch with the `error`, removing it across files.

File: pkg/controllers/awsloadbalancercontroller/rbac_clusterrole.go [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+		return err
+	}
+
+	if !changed {
+		return nil
+	}
+
+	updated := current.DeepCopy()
+	updated.Name = desired.Name
+	updated.Namespace = desired.Namespace
```
Comment (arjunrn) [inline pkg/controllers/awsloadbalancercontroller/rbac_clusterrole.go:67]: Setting name and namespace is not required because it's not possible. Also update with a new name and namespace will fail.

File: pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+		return err
+	}
+
+	if !changed {
+		return nil
+	}
+
+	updated := current.DeepCopy()
+	updated.Name = desired.Name
+	updated.Namespace = desired.Namespace
```
Comment (arjunrn) [inline pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go:67]: Setting name and namespace is not required because it's not possible. Also update with a new name and namespace will fail.

File: pkg/controllers/awsloadbalancercontroller/rbac_role.go [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+		return err
+	}
+
+	if !changed {
+		return nil
+	}
+
+	updated := current.DeepCopy()
+	updated.Name = desired.Name
+	updated.Namespace = desired.Namespace
```
Comment (arjunrn) [inline pkg/controllers/awsloadbalancercontroller/rbac_role.go:66]: Setting name and namespace is not required because it's not possible. Also update with a new name and namespace will fail.

File: pkg/controllers/awsloadbalancercontroller/rbac_rolebinding.go [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+	if err != nil {
+		return err
+	}
+
+	if !changed {
+		return nil
+	}
+	updated := current.DeepCopy()
+	updated.Name = desired.Name
+	updated.Namespace = desired.Namespace
```
Comment (arjunrn) [inline pkg/controllers/awsloadbalancercontroller/rbac_rolebinding.go:66]: Setting name and namespace is not required because it's not possible. Also update with a new name and namespace will fail.

File: pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+func isClusterRoleBindingChanged(current *rbacv1.ClusterRoleBinding, desired *rbacv1.ClusterRoleBinding) (bool, error) {
+	if current.Name != desired.Name {
+		return true, nil
+	}
+
+	if current.Namespace != desired.Namespace {
+		return true, nil
+	}
+
+	if !(reflect.DeepEqual(current.Subjects, desired.Subjects)) || !(reflect.DeepEqual(current.RoleRef, desired.RoleRef)) {
```
Comment (arjunrn) [inline pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go:87]: I'm afraid this will always fail because `Subject` has `APIGroup` field which has defaults which is not set in the desired.

Comment (arjunrn) [reply pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go:87]: Actually, not it will remain empty. So this is ok.

Comment (arjunrn) [reply pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go:87]: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#subject-v1-rbac-authorization-k8s-io

Comment (thejasn) [reply pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go:87]: It defaults to "" for type service accounts.

File: pkg/controllers/awsloadbalancercontroller/rbac_rolebinding.go (hunk 20 above)
Comment (thejasn) [reply pkg/controllers/awsloadbalancercontroller/rbac_rolebinding.go:66]: Addressed.

File: pkg/controllers/awsloadbalancercontroller/rbac_role.go (hunk 19 above)
Comment (thejasn) [reply pkg/controllers/awsloadbalancercontroller/rbac_role.go:66]: Addressed.

File: pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go (hunk 18 above)
Comment (thejasn) [reply pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go:67]: Addressed.

File: pkg/controllers/awsloadbalancercontroller/rbac_clusterrole.go (hunk 17 above)
Comment (thejasn) [reply pkg/controllers/awsloadbalancercontroller/rbac_clusterrole.go:67]: Addressed.

File: pkg/controllers/awsloadbalancercontroller/rbac_rolebinding.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (thejasn) [reply pkg/controllers/awsloadbalancercontroller/rbac_rolebinding.go:77]: Addressed.

File: pkg/controllers/awsloadbalancercontroller/rbac_role.go (hunk 15 above)
Comment (thejasn) [reply pkg/controllers/awsloadbalancercontroller/rbac_role.go:76]: Addressed.

File: pkg/controllers/awsloadbalancercontroller/rbac_role.go (hunk 14 above)
Comment (thejasn) [reply pkg/controllers/awsloadbalancercontroller/rbac_role.go:84]: Addressed.

File: pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go (hunk 13 above)
Comment (thejasn) [reply pkg/controllers/awsloadbalancercontroller/rbac_clusterrolebinding.go:85]: Addressed.
